File: src/platform/tauri/documentIo.ts
import { invokeNativeCommand } from './nativeCommands';

export interface FileSnapshotDto {
  mtimeMs: number | null;
  size: number | null;
}

export interface DocumentFileSessionDto {
  path: string;
  name: string;
  content: string;
  knownSnapshot: FileSnapshotDto | null;
}

export interface WriteDocumentFileInput {
  path: string;
  content: string;
  expectedSnapshot?: FileSnapshotDto | null;
  createNew?: boolean;
}

export interface DocumentWriteResultDto {
  path: string;
  snapshot: FileSnapshotDto;
}

export function getFileSnapshotNative(path: string) {
  return invokeNativeCommand<FileSnapshotDto>('get_file_snapshot', { path });
}

export function readDocumentFileNative(path: string) {
  return invokeNativeCommand<DocumentFileSessionDto>('read_document_file', { path });
}

export function writeDocumentFileNative(input: WriteDocumentFileInput) {
  return invokeNativeCommand<DocumentWriteResultDto>('write_document_file', { input });
}

File: src-tauri/src/commands/document_io.rs
use crate::domain::document_io::{
    self, DocumentFileSessionDto, DocumentWriteResult, FileSnapshotDto, WriteDocumentInput,
};
use crate::domain::error::PrismResult;

#[tauri::command]
pub fn get_file_snapshot(path: String) -> PrismResult<FileSnapshotDto> {
    document_io::get_file_snapshot(path)
}

#[tauri::command]
pub fn read_document_file(path: String) -> PrismResult<DocumentFileSessionDto> {
    document_io::read_document_file(path)
}

#[tauri::command]
pub fn write_document_file(input: WriteDocumentInput) -> PrismResult<DocumentWriteResult> {
    document_io::write_document_file(input)
}
